### PR TITLE
Fixes the oauth2 login.

### PR DESCRIPTION
**Pull request by alesstimec**
A bit of logic was missing to pass the account details through to where an oauth2 login provider is being created.

## Checklist

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [ x ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Manual QA with JIMM:
1) run a local JIMM instance
2) add a juju controller to JIMM
3) run `juju login <jimm dns>
4) inspect accounts.yaml to see the account type recorded as `oauth2-device`
5) run `juju models`, which should return an empty list, not an error

### CODE REVIEW

File: juju/api.go [hunk 1]
```diff
@@ -160,16 +160,18 @@ func NewAPIConnection(args NewAPIConnectionParams) (_ api.Connection, err error)
...
 			accountType = accountDetails.Type
 		}
-		if ok && !user.IsLocal() && apiInfo.Tag == nil && accountType != jujuclient.OAuth2DeviceFlowAccountDetailsType {
-			// We used macaroon auth to login; save the username
-			// that we've logged in as.
-			accountDetails = &jujuclient.AccountDetails{
-				Type:            jujuclient.UserPassAccountDetailsType,
-				User:            user.Id(),
-				LastKnownAccess: st.ControllerAccess(),
+		if accountType == "" || accountType == jujuclient.UserPassAccountDetailsType {
```
Comment (wallyworld) [inline juju/api.go:163]: We can look to make this cleaner as well; maybe it can call the same function

File: cmd/modelcmd/base_test.go [hunk 2]
```diff
@@ -407,3 +408,58 @@ func addCookie(c *gc.C, jar http.CookieJar, mac *macaroon.Macaroon, url *url.URL
 	cookie.Expires = time.Now().Add(time.Hour) // only persistent cookies are stored
 	jar.SetCookies(url, []*http.Cookie{cookie})
 }
+
+func (s *BaseCommandSuite) TestNewAPIRootWithOAuthLoginProvider(c *gc.C) {
```
Comment (wallyworld) [inline cmd/modelcmd/base_test.go:412]: If we move the code to a function as suggested, that function can be tested in isolation and avoid all this.
We don't want to use reflect in tests - if we need access to unexported items we'd write the test in the same package as the code.

File: go.mod [hunk 3]
```diff
@@ -29,6 +29,7 @@ require (
 	github.com/coreos/go-systemd/v22 v22.5.0
 	github.com/docker/distribution v2.8.3+incompatible
 	github.com/dustin/go-humanize v1.0.1
+	github.com/frankban/quicktest v1.14.6
```
Comment (wallyworld) [inline go.mod:32]: Delete me

File: cmd/modelcmd/base.go [hunk 4]
```diff
@@ -230,20 +230,22 @@ func (c *CommandBase) NewAPIRootWithDialOpts(
...
-			if len(accountDetails.Macaroons) == 0 {
-				accountDetails = &jujuclient.AccountDetails{}
-			} else {
-				// If the account has macaroon set, use those to login
-				// to avoid an unnecessary auth round trip.
-				// Used for embedded commands.
-				accountDetails = &jujuclient.AccountDetails{
-					User:      u.Id(),
-					Macaroons: accountDetails.Macaroons,
+	if accountDetails.Type == "" || accountDetails.Type == jujuclient.UserPassAccountDetailsType {
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply cmd/modelcmd/base.go]: done

File: juju/api.go (hunk 1 above)
Comment (alesstimec) [reply juju/api.go:177]: done.. extracted account detail processing into its own function

File: cmd/modelcmd/base_test.go (hunk 2 above)
Comment (alesstimec) [reply cmd/modelcmd/base_test.go]: done

File: go.mod (hunk 3 above)
Comment (alesstimec) [reply go.mod]: done

**Review by wallyworld — APPROVED** [review]

**Pull request merged**